File: ruby/grains/grains.rb
# frozen_string_literal: true

class Grains
  MIN_SQUARE = 1
  MAX_SQUARE = 64

  private_constant :MIN_SQUARE, :MAX_SQUARE

  def self.square(square_number)
    raise BoardSquareError unless valid_board_square?(square_number)

    1 << (square_number - 1)
  end

  def self.total
    # as each square doubles, the total amount of grains on the board is
    # always one less than the grains that would be in the next square
    square(MAX_SQUARE) * 2 - 1
  end

  def self.valid_board_square?(square_number)
    MIN_SQUARE <= square_number && square_number <= MAX_SQUARE
  end
end

class BoardSquareError < ArgumentError
  def message
    'not a valid board square number'
  end
end

File: go/grains/grains.go
package grains

import (
	"fmt"
)

func Square(n int) (uint64, error) {
	if n > 64 || n < 1 {
		return 0, fmt.Errorf("invalid square '%d'", n)
	}

	return uint64(1 << (n - 1)), nil
}

func Total() uint64 {
	return ^uint64(0)
}
